pkg/devicetwin/dtcontext/dtcontext_test.go

```diff
@@ -0,0 +1,451 @@
+package dtcontext
+
+import (
+	"reflect"
+	"sync"
+	"testing"
+
+	"github.com/kubeedge/kubeedge/beehive/pkg/common/config"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/context"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/pkg/common/util"
+	"github.com/kubeedge/kubeedge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/pkg/devicetwin/dttype"
+)
+
+// TestInitDTContext is function to test InitDTContext().
+func TestInitDTContext(t *testing.T) {
+	util.LoadConfig()
+	nodeID, err := config.CONFIG.GetValue("edgehub.controller.node-id").ToString()
+	if err != nil {
+		t.Errorf("Error in getting node id %v", err)
+		return
+	}
+	tests := []struct {
+		name    string
+		context *context.Context
+		want    *DTContext
+		wantErr error
+	}{
+		{
+			name:    "ActualContextArgumentTest",
+			context: context.GetContext(context.MsgCtxTypeChannel),
+			want: &DTContext{
+				GroupID:        "",
+				NodeID:         nodeID,
+				CommChan:       make(map[string]chan interface{}, 1),
+				ConfirmChan:    make(chan interface{}, 1),
+				ConfirmMap:     &sync.Map{},
+				ModulesHealth:  &sync.Map{},
+				ModulesContext: context.GetContext(context.MsgCtxTypeChannel),
+				DeviceList:     &sync.Map{},
+				DeviceMutex:    &sync.Map{},
+				Mutex:          &sync.Mutex{},
+				State:          dtcommon.Disconnected,
+			},
+			wantErr: nil,
+		},
+		{
+			name:    "EmptyContextArgumentTest",
+			context: context.GetContext("test"),
+			want: &DTContext{
+				GroupID:        "",
+				NodeID:         nodeID,
+				CommChan:       make(map[string]chan interface{}),
+				ConfirmChan:    make(chan interface{}, 1),
+				ConfirmMap:     &sync.Map{},
+				ModulesHealth:  &sync.Map{},
+				ModulesContext: context.GetContext("test"),
+				DeviceList:     &sync.Map{},
+				DeviceMutex:    &sync.Map{},
+				Mutex:          &sync.Mutex{},
+				State:          dtcommon.Disconnected,
+			},
+			wantErr: nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := InitDTContext(test.context)
+			if err != test.wantErr {
+				t.Errorf("InitDTContext() error = %v, wantErr %v", err, test.wantErr)
+			}
+			if !reflect.DeepEqual(got.GroupID, test.want.GroupID) {
+				t.Errorf("GroupID = %v, want = %v", got.GroupID, test.want.GroupID)
+			}
+			if !reflect.DeepEqual(got.NodeID, test.want.NodeID) {
+				t.Errorf("NodeID = %v, want = %v", got.NodeID, test.want.NodeID)
+			}
+			if !reflect.DeepEqual(got.CommChan, test.want.CommChan) {
+				t.Errorf("CommunicationChannel = %v, want =%v", got.CommChan, test.want.CommChan)
+			}
+			if reflect.TypeOf(got.ConfirmChan) != reflect.TypeOf(test.want.ConfirmChan) {
+				t.Errorf("ConfirmChannel = %v, want =%v", reflect.TypeOf(got.ConfirmChan), reflect.TypeOf(test.want.ConfirmChan))
```

Comment:
the initDTContext method creates a confirmChan of size 1000, but in our tests , the expected size is 1, so ideally this test should have failed and wantError won't be nil. can we have one test case with size as 1000 which passes and one with any other size which fails ?

Reply:
we are checking type here only , we need to test actual chan size.